src/devices/bus/heathzenith/h8/h8bus.h

```diff
@@ -0,0 +1,399 @@
+
+	template <typename T>
+	void set_h8bus_slot(T &&tag, const char *slottag)
+	{
+		m_h8bus.set_tag(std::forward<T>(tag));
+		m_h8bus_slottag = slottag;
+	}
+
+protected:
+	h8bus_slot_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock);
+
+	virtual void device_resolve_objects() override ATTR_COLD;
+	virtual void device_start() override ATTR_COLD;
+
+	required_device<h8bus_device> m_h8bus;
+	const char *m_h8bus_slottag;
+};
+
+DECLARE_DEVICE_TYPE(H8BUS_SLOT, h8bus_slot_device)
+
+
+class h8bus_device : public device_t, public device_memory_interface
+{
+	friend class h8bus_slot_device;
+	friend class device_h8bus_card_interface;
+	friend class device_h8bus_p1_card_interface;
+	friend class device_h8bus_p2_card_interface;
+
+public:
+	h8bus_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+	~h8bus_device();
+
+protected:
+	h8bus_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock);
+
+	virtual void interface_pre_start() override;
+	virtual void device_start() override ATTR_COLD;
+	virtual void device_reset() override ATTR_COLD;
+
+	// device_memory_interface overrides
+	virtual space_config_vector memory_space_config() const override;
+
+	void add_h8bus_card(device_h8bus_card_interface &card);
+	void set_int1_line(int state);
+	void set_int2_line(int state);
+	void set_int3_line(int state);
+	void set_int4_line(int state);
+	void set_int5_line(int state);
+	void set_int6_line(int state);
+	void set_int7_line(int state);
+	void set_p201_inte(int state);
+	void set_p201_reset(int state);
+	void set_p201_int1(int state);
+	void set_p201_int2(int state);
+	void set_reset_line(int state);
+	void set_hold_line(int state);
+	void set_disable_rom_line(int state);
+	void set_m1_line(int state);
+
+	void mem_map(address_map &map);
+	void io_map(address_map &map);
+
+private:
+	address_space_config const m_mem_config;
+	address_space_config const m_io_config;
+
+	std::vector<std::reference_wrapper<device_h8bus_card_interface>> m_device_list;
+
+	device_h8bus_p1_card_interface *m_p1_card;
+	device_h8bus_p2_card_interface *m_p2_card;
+};
+
+
+inline void device_h8bus_card_interface::set_slot_int1(int state)
+{
+	h8bus().set_int1_line(state);
+}
+
+inline void device_h8bus_card_interface::set_slot_int2(int state)
+{
+	h8bus().set_int2_line(state);
+}
+
+inline void device_h8bus_card_interface::set_slot_int3(int state)
+{
+	h8bus().set_int3_line(state);
+}
+
+inline void device_h8bus_card_interface::set_slot_int4(int state)
+{
+	h8bus().set_int4_line(state);
+}
+
+inline void device_h8bus_card_interface::set_slot_int5(int state)
+{
+	h8bus().set_int5_line(state);
+}
+
+inline void device_h8bus_card_interface::set_slot_int6(int state)
+{
+	h8bus().set_int6_line(state);
+}
+
+inline void device_h8bus_card_interface::set_slot_int7(int state)
+{
+	h8bus().set_int7_line(state);
+}
```

Comment:
This naïve approach to implementing “party line” signals causes issues if multiple cards assert the same signal simultaneously.  Consider what would happen with the code as-is if two cards both assert INT2:

* Card 1 asserts INT2, bus sees INT2 asserted (correct)
* Card 2 asserts INT2, bus still sees INT2 asserted (correct)
* Card 1 deasserts INT2, bus sees INT2 deasserted (incorrect, card 2 still has it asserted)
* Card 2 deasserts INT2, bus sees INT2 deasserted (correct)

This is why the INTELLEC 4 bus (src/devices/bus/intellec4) keeps track of which cards have which lines asserted in bit flags.  It also facilitates implementing cards that react when *any other card* asserts a line.  I’d suggest doing something similar.

Reply:
I'm not aware of any Heath software that would properly handle shared interrupts. Would it be ok, to add a TODO in this PR about this issue, and address it in a followup PR?